### PR TITLE
Remove state_dict from AveragedModel and use buffers instead

### LINKED ISSUE #66686: Improve use_state_dict in AveragedModel
use_state_dict was added in AveragedModel in https://github.com/pytorch/pytorch/pull/65495 and https://github.com/pytorch/pytorch/pull/65921.
This needs to be further improved as described [here](https://github.com/pytorch/pytorch/pull/65755#issuecomment-937983130). This issue has been created to keep track of this work item.

cc @vincentqb @jbschlosser @albanD @datumbox

### PR DESCRIPTION
**Pull request by prabhat00155**
Fixes [#66686](https://github.com/pytorch/pytorch/issues/66686)

### CODE REVIEW

File: torch/optim/swa_utils.py [hunk 1]
```diff
@@ -26,8 +28,11 @@ class AveragedModel(Module):
             :class:`AveragedModel` parameter, the current value of :attr:`model`
             parameter and the number of models already averaged; if None,
             equally weighted average is used (default: None)
-        mode (str, optional): whether to use ``'parameters'`` or ``'state_dict'`` for update
-            (default: ``'parameters'``)
+        use_buffers (bool, optional): If the model has layers that contain
+            buffers(like BN), passing `use_buffers=True` would use buffers
```
Comment (albanD) [inline torch/optim/swa_utils.py:32]: nit space

File: torch/optim/swa_utils.py [hunk 2]
```diff
@@ -26,8 +28,11 @@ class AveragedModel(Module):
...
+            along with parameters to perform the update. This has shown to
```
Comment (albanD) [inline torch/optim/swa_utils.py:33]: nit: "perform the update" is not super clear? Maybe just say that buffers also get averaged alongside parameters.

Comment (datumbox) [reply torch/optim/swa_utils.py]: @albanD You caught us in the middle of updating the documentation. :) Could you have another look?

Comment (albanD) [reply torch/optim/swa_utils.py]: Sure no problem, it is a draft so I was early haha

File: test/test_optim.py [hunk 3]
```diff
@@ -2486,7 +2486,7 @@ def test_averaged_model_exponential_use_state_dict(self):
 
         def avg_fn(p_avg, p, n_avg):
             return alpha * p_avg + (1 - alpha) * p
-        averaged_dnn = AveragedModel(dnn, avg_fn=avg_fn, mode='state_dict')
+        averaged_dnn = AveragedModel(dnn, avg_fn=avg_fn, use_buffers=True)
```
[the remark this answers is not in the document]
Comment (prabhat00155) [reply test/test_optim.py:2491]: Updated the test.

File: test/test_optim.py [hunk 4]
```diff
@@ -2486,13 +2488,14 @@ def test_averaged_model_exponential_use_state_dict(self):
 
         def avg_fn(p_avg, p, n_avg):
             return alpha * p_avg + (1 - alpha) * p
-        averaged_dnn = AveragedModel(dnn, avg_fn=avg_fn, mode='state_dict')
-        averaged_params = [torch.zeros_like(param) for param in dnn.state_dict().values()
+        averaged_dnn = AveragedModel(dnn, avg_fn=avg_fn, use_buffers=True)
```
Comment (albanD) [inline test/test_optim.py:2491]: Do we have another test that covers the `use_buffers=False` case?
If not we should update this. You can use `@parametrize` decorator for the test to do that nicely.

**Review by albanD — COMMENTED** [review]
Looks good. Small test update and this is good to go I think.

File: test/test_optim.py (hunk 4 above)
Comment (prabhat00155) [reply test/test_optim.py:2491]: By default, `use_buffers` is `False`. This case is already covered by https://github.com/pytorch/pytorch/blob/0fdb90da5e1e7ec8f6becbdeecb84a3d855268ad/test/test_optim.py#L2450.

Comment (albanD) [reply test/test_optim.py:2491]: Right, I guess we could deduplicate code. But this update sounds good.

**Review by albanD — APPROVED** [review]
Thanks for the update!

**Review by jbschlosser — APPROVED** [review]
Seems good to me

**Pull request closed without merging**